File: services/us_factors/analyst.py
"""美股分析师因子: US_ANALYST_RATING, US_ANALYST_COVERAGE"""

import logging

import numpy as np
import pandas as pd

from services.config import LOG_LEVEL
from services.us_factors.base import USFactorBase

logger = logging.getLogger(__name__)
logger.setLevel(LOG_LEVEL)

# 评级映射（越高越看好）
_RATING_MAP = {
    "Strong Buy": 5.0, "Buy": 4.0, "Outperform": 4.0, "Overweight": 4.0,
    "Market Perform": 3.0, "Hold": 3.0, "Neutral": 3.0, "Equal-Weight": 3.0,
    "Sector Perform": 3.0, "In-Line": 3.0, "Peer Perform": 3.0,
    "Underperform": 2.0, "Underweight": 2.0, "Reduce": 2.0,
    "Sell": 1.0, "Strong Sell": 1.0,
}

# 回看窗口
_LOOKBACK_DAYS = 120


class USAnalystRating(USFactorBase):
    """Analyst Rating: mean rating score over trailing window"""
    name = "US_ANALYST_RATING"
    description = "分析师共识评级"

    def compute(self, date: str, universe: pd.DataFrame) -> pd.DataFrame:
        tickers = universe["ticker"].tolist()
        date_ts = pd.to_datetime(date)
        start_ts = date_ts - pd.Timedelta(days=_LOOKBACK_DAYS)

        # 从预加载数据获取
        bulk_ar = self._static_cache.get("_bulk_analyst")
        if bulk_ar is not None and not bulk_ar.empty:
            mask = (bulk_ar["date"] >= start_ts) & (bulk_ar["date"] <= date_ts)
            df = bulk_ar[mask].copy()
        else:
            df = self.db.query(
                "SELECT ticker, date, rating FROM us_analyst_recommendation "
                "WHERE date >= :start AND date <= :end",
                params={"start": start_ts.strftime("%Y-%m-%d"), "end": date},
            )

        if df.empty:
            logger.debug("USAnalystRating.compute: 无分析师推荐数据")
            return pd.DataFrame(columns=["ticker", "factor_value"])

        if tickers:
            df = df[df["ticker"].isin(tickers)]

        # 映射评级到数值
        df["score"] = df["rating"].str.strip().map(_RATING_MAP)
        df = df.dropna(subset=["score"])

        if df.empty:
            logger.debug("USAnalystRating.compute: 评级映射后无有效数据")
            return pd.DataFrame(columns=["ticker", "factor_value"])

        result = df.groupby("ticker")["score"].mean().reset_index()
        result.columns = ["ticker", "factor_value"]
        return result


class USAnalystCoverage(USFactorBase):
    """Analyst Coverage: log(1 + distinct analyst firms) in trailing window"""
    name = "US_ANALYST_COVERAGE"
    description = "分析师覆盖度"

    def compute(self, date: str, universe: pd.DataFrame) -> pd.DataFrame:
        tickers = universe["ticker"].tolist()
        date_ts = pd.to_datetime(date)
        start_ts = date_ts - pd.Timedelta(days=_LOOKBACK_DAYS)

        bulk_ar = self._static_cache.get("_bulk_analyst")
        if bulk_ar is not None and not bulk_ar.empty:
            mask = (bulk_ar["date"] >= start_ts) & (bulk_ar["date"] <= date_ts)
            df = bulk_ar[mask].copy()
        else:
            df = self.db.query(
                "SELECT ticker, analyst_company FROM us_analyst_recommendation "
                "WHERE date >= :start AND date <= :end",
                params={"start": start_ts.strftime("%Y-%m-%d"), "end": date},
            )

        if df.empty:
            logger.debug("USAnalystCoverage.compute: 无分析师推荐数据")
            return pd.DataFrame(columns=["ticker", "factor_value"])

        if tickers:
            df = df[df["ticker"].isin(tickers)]

        result = df.groupby("ticker")["analyst_company"].nunique().reset_index()
        result.columns = ["ticker", "n_firms"]
        result["factor_value"] = np.log1p(result["n_firms"])
        return result[["ticker", "factor_value"]]

File: quant-engine/crates/factors/src/analyst/analyst.rs
//! Analyst factors: US_ANALYST_RATING, US_ANALYST_COVERAGE

use std::collections::HashSet;

use quant_core::types::{Date, FactorResult};
use quant_data::cache::DataCache;

use crate::registry::Factor;

const LOOKBACK_DAYS: i64 = 120;

fn rating_score(grade: &str) -> Option<f64> {
    match grade.trim() {
        "Strong Buy" => Some(5.0),
        "Buy" | "Outperform" | "Overweight" => Some(4.0),
        "Market Perform" | "Hold" | "Neutral" | "Equal-Weight"
        | "Sector Perform" | "In-Line" | "Peer Perform" => Some(3.0),
        "Underperform" | "Underweight" | "Reduce" => Some(2.0),
        "Sell" | "Strong Sell" => Some(1.0),
        _ => None,
    }
}

/// Analyst Rating: mean rating score over trailing window
pub struct UsAnalystRating;
inventory::submit! { &UsAnalystRating as &dyn Factor }

impl Factor for UsAnalystRating {
    fn name(&self) -> &'static str { "US_ANALYST_RATING" }
    fn category(&self) -> &'static str { "analyst" }
    fn inherent_direction(&self) -> i8 { 0 }
    fn ic_window_months(&self) -> u32 { 12 }

    fn compute(&self, date: Date, cache: &DataCache) -> FactorResult {
        let start = date - chrono::Duration::days(LOOKBACK_DAYS);
        let mut result = FactorResult::default();

        for (&tid, recs) in &cache.analyst_recs {
            let scores: Vec<f64> = recs
                .iter()
                .filter(|r| r.date >= start && r.date <= date)
                .filter_map(|r| rating_score(&r.new_grade))
                .collect();
            if !scores.is_empty() {
                let mean = scores.iter().sum::<f64>() / scores.len() as f64;
                result.insert(tid, mean);
            }
        }
        result
    }
}

/// Analyst Coverage: log(1 + distinct analyst firms) in trailing window
pub struct UsAnalystCoverage;
inventory::submit! { &UsAnalystCoverage as &dyn Factor }

impl Factor for UsAnalystCoverage {
    fn name(&self) -> &'static str { "US_ANALYST_COVERAGE" }
    fn category(&self) -> &'static str { "analyst" }
    fn inherent_direction(&self) -> i8 { 0 }
    fn ic_window_months(&self) -> u32 { 12 }

    fn compute(&self, date: Date, cache: &DataCache) -> FactorResult {
        let start = date - chrono::Duration::days(LOOKBACK_DAYS);
        let mut result = FactorResult::default();

        for (&tid, recs) in &cache.analyst_recs {
            let firms: HashSet<&str> = recs
                .iter()
                .filter(|r| r.date >= start && r.date <= date)
                .map(|r| r.grading_company.as_str())
                .filter(|s| !s.is_empty())
                .collect();
            if !firms.is_empty() {
                result.insert(tid, (1.0 + firms.len() as f64).ln());
            }
        }
        result
    }
}

/// Insider Net Buy: net insider purchases / market cap over trailing 90 days
pub struct InsiderNetBuy;
inventory::submit! { &InsiderNetBuy as &dyn Factor }

impl Factor for InsiderNetBuy {
    fn name(&self) -> &'static str { "INSIDER_NET_BUY" }
    fn category(&self) -> &'static str { "analyst" }
    fn inherent_direction(&self) -> i8 { 0 }
    fn ic_window_months(&self) -> u32 { 12 }

    fn compute(&self, date: Date, cache: &DataCache) -> FactorResult {
        let start = date - chrono::Duration::days(90);
        let mut result = FactorResult::default();

        for (&tid, trades) in &cache.insider_trades {
            let net_value: f64 = trades
                .iter()
                .filter(|t| t.filing_date >= start && t.filing_date <= date)
                .map(|t| {
                    let val = t.securities_transacted * t.price;
                    if t.is_acquisition { val } else { -val }
                })
                .filter(|v| v.is_finite())
                .sum();

            if net_value.abs() < 1e-6 {
                continue;
            }

            let mktcap = match cache.get_market_cap(tid, date) {
                Some(m) if m > 0.0 => m,
                _ => continue,
            };

            let ratio = net_value / mktcap;
            if ratio.is_finite() {
                result.insert(tid, ratio);
            }
        }
        result
    }
}
